src/main/java/net/minecraftforge/server/command/TagsCommand.java

```diff
@@ -0,0 +1,266 @@
+import net.minecraft.core.HolderSet;
+import net.minecraft.core.Registry;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.ComponentUtils;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+import net.minecraft.network.chat.TextComponent;
+import net.minecraft.network.chat.TranslatableComponent;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.tags.TagKey;
+import net.minecraft.util.Mth;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The {@code /forge tags} command for listing a registry's tags, getting the elements of tags, and querying the tags of a
+ * registry object.
+ *
+ * <p>Each command is paginated, showing {@value PAGE_SIZE} entries at a time. When there are more than 0 entries,
+ * the text indicating the amount of entries is highlighted and can be clicked to copy the list of all entries (across
+ * all pages) to the clipboard. (This is reflected by the use of green text in brackets, mimicking the clickable
+ * coordinates in the {@code /locate} command's message)</p>
+ *
+ * <p>The command has three subcommands:</p>
+ * <ul>
+ *     <li>{@code /forge tags &lt;registry> list [page]} - Lists all available tags in the given registry.</li>
+ *     <li>{@code /forge tags &lt;registry> get &lt;tag> [page]} - Gets all elements of the given tag in the given registry.</li>
+ *     <li>{@code /forge tags &lt;registry> query &lt;element> [page]} - Queries for all tags in the given registry which
+ *     contain the given registry object.</li>
+ * </ul>
+ */
+class TagsCommand
+{
+    private static final long PAGE_SIZE = 8;
+    private static final ResourceKey<Registry<Registry<?>>> ROOT_REGISTRY_KEY =
+            ResourceKey.createRegistryKey(new ResourceLocation("root"));
+
+    private static final DynamicCommandExceptionType UNKNOWN_REGISTRY = new DynamicCommandExceptionType(key ->
+            new TranslatableComponent("Unknown registry '%s'", key));
+    private static final Dynamic2CommandExceptionType UNKNOWN_TAG = new Dynamic2CommandExceptionType((tag, registry) ->
+            new TranslatableComponent("Unknown tag '%s' in registry '%s'", tag, registry));
+    private static final Dynamic2CommandExceptionType UNKNOWN_ELEMENT = new Dynamic2CommandExceptionType((tag, registry) ->
+            new TranslatableComponent("Unknown element '%s' in registry '%s'", tag, registry));
+
+    public static ArgumentBuilder<CommandSourceStack, ?> register()
+    {
+        /*
+         * /forge tags <registry> list [page]
+         * /forge tags <registry> get <tag> [page]
+         * /forge tags <registry> query <element> [page]
+         */
+        return Commands.literal("tags")
```

Comment:
I don't see a major reason not to allow this, but do we want this command to be accessible by everyone by default (like it is now)? Or do we want to potentially require the sender to be an op